### PR TITLE
[SPARK-33454][INFRA] Add GitHub Action job for Hadoop 2

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to protect `Hadoop 2.x` profile compilation in Apache Spark 3.1+.

### Why are the changes needed?

Since Apache Spark 3.1+ switch our default profile to Hadoop 3, we had better prevent at least compilation error with `Hadoop 2.x` profile at the PR review phase. Although this is an additional workload, it will finish quickly because it's compilation only.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass the GitHub Action.
- This should be merged after https://github.com/apache/spark/pull/30375 .

### CODE REVIEW

File: .github/workflows/build_and_test.yml
Comment (dongjoon-hyun) [inline .github/workflows/build_and_test.yml:424]: We can upgrade from Hadoop 2.7.4 to Hadoop 2.10.x. So, the job name is only mentioning the major version.

**Comment by mridulm** [comment]
Thanks for making this change @dongjoon-hyun !
This applies only to github actions right ? I assumed that currently the procedure was to wait for either jenkins or github actions to pass before merge.
With this change, a 2.x issue would be caught only by github actions, not jenkins.

Or does jenkins leverage the same ?

**Comment by dongjoon-hyun** [comment]
Ya. That's true. However, `GitHub Action` is very fast and `Hadoop 2` build compilation error is easily detected. I means the result comes in 15 minutes and it's hard to consider as a flaky job.

For Jenkins, IIUC, it's difficult to apply in PR Builder. Currently, all PR Builders (PRBuilder, K8s IT PRBuilder) is focusing on the default profile because the runs are too heavy. You may ask Shane for that. He knows better than me about the capacity and new PR Builder for Hadoop 2.

**Comment by dongjoon-hyun** [comment]
BTW, I have only read-only access privilege to see Jenkins configuration.

**Comment by mridulm** [comment]
+CC @shaneknapp 
Will it be possible to do the same for jenkins ?
It will ensure consistency and no accidental merges.

**Comment by mridulm** [comment]
>  I means the result comes in 15 minutes and it's hard to consider it as a flaky job.
I am concerned sometimes we might just not see it - given it is listed at the bottom of the list (if jenkins gives green).
If we cant do it in both, atleast reordering so that build failures are called out first would be good. Thoughts @dongjoon-hyun ?

**Comment by dongjoon-hyun** [comment]
Sure! Given the importance of Hadoop 2, it sounds good to me.

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:22]: Yeah, we should get rid of Maven cache which is not used. Nice.

**Comment by HyukjinKwon** [comment]
There looks a legitimate failure but I like this idea +1.

**Comment by dongjoon-hyun** [comment]
Yes, @HyukjinKwon . The failure is the motive of this PR. We should prevent that happens again.


Screen Shot 2020-11-15 at 6 32 40 PM

**Comment by dongjoon-hyun** [comment]
Since #30375 is merged now, I retriggered GitHub Action.

cc @mridulm , @HyukjinKwon , @wangyum

**Review by HyukjinKwon — APPROVED** [review]

File: .github/workflows/build_and_test.yml
Comment (viirya) [inline .github/workflows/build_and_test.yml:34]: Does Java version cause any difference here?

**Review by viirya — COMMENTED** [review]
Looks okay, just one question.

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml:34]: Actually, maybe we should stick to Java 8 because that's the Java version being used in the releases.

Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:34]: No, @viirya . Apache Spark 3.0 already declared [Hadoop 2.7 + Hive 2.3 on JDK11](https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-branch-3.0-test-sbt-hadoop-2.7-hive-2.3/).
> Does Java version cause any difference here?

Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:34]: @HyukjinKwon . It will be okay~ This is only for compilation test for API instead of running real tests.

After we make `branch-3.1`, Jenkins will be created for all release profiles, too.

**Comment by dongjoon-hyun** [comment]
It seems that there exists some delay on GitHub Action. I retriggered GitHub Action because the previous run fails with compilation error again.

**Comment by dongjoon-hyun** [comment]
BTW, @HyukjinKwon . Do you still want to use Java 8 in this PR?

**Review by wangyum — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you, @wangyum . I'm looking at both this PR and vanilla master branch, too.

**Comment by dongjoon-hyun** [comment]
It's weird. The new code is not consumed at the GitHub Action re-trigger. I'll rebase this to the master~
```
[error] /home/runner/work/spark/spark/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:320:52: type mismatch;
706
[error]  found   : Long
707
[error]  required: Int
708
[error]         Resource.newInstance(resourcesWithDefaults.totalMemMiB, resourcesWithDefaults.cores)
709
[error]                                                    ^
710
[error] one error found
```

**Comment by wangyum** [comment]
Thank you @dongjoon-hyun.

**Review by viirya — APPROVED** [review]

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml]: Sure, I guess it's fine.

**Comment by dongjoon-hyun** [comment]
Sorry for the rebasing after your approval, @HyukjinKwon , @wangyum , @viirya . It was inevitable to bring the latest master branch. (I didn't notice that GitHub Action doesn't use the latest master.)

Also, I switched to Java 8 according to @HyukjinKwon 's advice.

**Comment by dongjoon-hyun** [comment]
Now, the new job is green.
Screen Shot 2020-11-15 at 9 46 58 PM

**Comment by dongjoon-hyun** [comment]
Could you approve once more please?

**Review by viirya — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by dongjoon-hyun** [comment]
Yay! Thank you, @viirya and @HyukjinKwon ! 😄

**Comment by wangyum** [comment]
Another related question, do you think we should test Java 14 by Github Action?

**Comment by HyukjinKwon** [comment]
There's a discussion thread in the mailing list about JDK 14 FYI: http://apache-spark-developers-list.1001551.n3.nabble.com/Spark-on-JDK-14-td30348.html

**Comment by dongjoon-hyun** [comment]
Although it's important to track upstream, Java 14 is already EOL.
Only Java 8 / Java 11 / Java 17 (September 2021) are LTS.

**Pull request closed without merging**
